### ISSUE TITLE
Elasticsearch nodes run into OOM during sustained ThreadPoolRejections

### ISSUE THREAD

**Issue by mahdibh** [elastic/elasticsearch#18230]
**Elasticsearch version**: 1.7.3
**JVM version**: Java HotSpot(TM) 64-Bit Server VM (build 25.66-b17, mixed mode)
**OS version**: Linux ip-XXXXXXX.ec2.internal 4.1.13-19.30.amzn1.x86_64 #1 SMP Fri Dec 11 03:42:10 -- UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
**Description of the problem including expected versus actual behavior**:
3 data nodes (7GB heap), 3 dedicated master nodes.
1 index with 128 primary shards, 2 replicas. Index has 600k documents.
Default configuration, with reduced search thread pool size (from default of 13 to 5 threads). We also changed the search queue size from 1000 to 999 (to verify that we could issue cluster level settings updates during the test).

**Steps to reproduce**:
We were hitting this issue in production and were able to successfully reproduce it in a staging environment by doing the following:

 1. spin off 1000 threads running the following query concurrently (same query across all threads) in a tight loop

`POST index_name/object/_search
{
  "size" : 5000,
  "query" : {
    "filtered" : {
      "query" : {
        "filtered" : {
          "query" : {
            "query_string" : {
              "query" : "(field1:value1 AND NOT field2:_PREFIX_*)",
              "lowercase_expanded_terms" : false
            }
          },
          "filter" : {
            "and" : {
              "filters" : [ {
                "term" : {
                  "field3" : "value2"
                }
              }, {
                "term" : {
                  "type" : "MTS"
                }
              }, {
                "or" : {
                  "filters" : [ {
                    "term" : {
                      "active" : true
                    }
                  }, {
                    "and" : {
                      "filters" : [ {
                        "term" : {
                          "active" : false
                        }
                      }, {
                        "range" : {
                          "lastActive" : {
                            "from" : 0,
                            "to" : null,
                            "include_lower" : true,
                            "include_upper" : true
                          },
                          "_cache" : false
                        }
                      } ]
                    }
                  } ]
                }
              } ]
            }
          }
        }
      },
      "filter" : {
        "missing" : {
          "field" : "_deletedOnMs",
          "null_value" : true,
          "existence" : true
        }
      }
    }
  },
  "version" : false,
  "_source" : {
    "includes" : [ ],
    "excludes" : [ "_props" ]
  },
  "sort" : [ {
    "createdOnMs" : {
      "order" : "asc"
    }
  } ]
}`

 2. those queries will rapidly fill up the search queue and we get thread pool rejections.
 3. after 5mn, all 3 data nodes get into a zombie state (ie, OOM'ed) and get kicked out of the cluster

**Provide logs (if relevant)**:
thread dumps
https://gist.github.com/mahdibh/025d7a909475c43f9154e661c3ef839f
https://gist.github.com/mahdibh/b89a1d437467339a5c30eb416a96cfb5

node log
https://gist.github.com/mahdibh/320b515788400fb1560f2da1b1f2897f

**Notes**
All nodes in the cluster run into the same issue. We took a heap dump of one of the nodes when it was in this state. 41% of the shallow size of the heap is consumed by byte arrays. long[] comes next with 16% and then char[] with 8%. We could privately share the heap dump if it helps figuring out what's going on.

When this happens, the clusters becomes useless. Only the master nodes respond to API calls. The head plugin shows a blank list of nodes.

We can easily reproduce this internally, if there is anything we can do to provide more details, please let us know.

**Comment by clintongormley** [comment]
Hi @mahdibh 

We have added eg the in-flight requests circuit breaker (https://github.com/elastic/elasticsearch/pull/17133) which may or may not help in this situation.  It depends exactly where memory is being consumed here.  eg the `requests` circuit breaker is applied during the query phase, but not during the fetch phase and not on the coordinating node.  So large `_source` fields might consume too much memory and result in an OOM.

Giving us access to the heap dump would be helpful.  Also, would you mind trying this same test out on 5.0.0-alpha2?

We may be able to find other safeguards (see https://github.com/elastic/elasticsearch/issues/11511) that will help us in these situations

**Comment by mahdibh** [comment]
Hi @clintongormley 

We will try to repro this without requesting _source and see if that helps. Our document _source is usually 10kb. Not sure if that actually will matter here since we have only 5 search threads per node.

Would the number of shards (128) coupled with the fact that we are retrieving 5000 records (sorted) contribute to the memory bloat when the system isn't able to keep up ? My understanding is that ES will have to sort and return the top 5000 docs on each shard to the coordinating node which will have to sort across 128x5000=640k docs. Our sort field has doc values enabled and it's a long timestamp.

Also, would the ThreadPoolRejection exceptions cause extra allocations (since there is an error per shard). When the system is overloaded, it's likely all shards will reject and the response object will contain 128 shard failure messages.

I'll follow up regarding the heapdump and trying this out on 5.0.0-alpha2

**Comment by mahdibh** [comment]
I tried removing the sort in the query, reducing the number of results fetched from 5k to 100, not fetching _source and using no regex in the query but still no luck. I can get the 3 data nodes to OOM in under 3 minutes of sustained load.

I then tried to use the same index but with 8 primary shards instead of 128 (+2 replica). In that case, I couldn't get the cluster to crash (nodes disappeared from the cluster temporarily, but then came back once I stopped the load).

I've also noticed that the cluster held-up more when I disabled the logging of ThreadPoolRejectionExceptions on the ES side PUT _cluster/settings" -d'{"transient" : {"logger.action.search.type" : "ERROR"}}'). Those were logged at a peak rate of 10k/second. No wonder that would consume CPU. I'm not sure why stack traces are logged for those exceptions by default, seems like shooting oneself in the foot when ES is trying to shed itself from further load.

@clintongormley any luck finding anything suspicious in the heap dump ?

**Comment by mahdibh** [comment]
More data points:
We have upgraded to 1.7.5, but still run into the same issue.

We switched our query traffic from using the data nodes directly to just using client nodes only instead. This shifted the OOM to the client nodes (which is good, since those are way much cheaper to restart). I'll share the heap dump we got on the client nodes.

**Comment by s1monw** [comment]
> Would the number of shards (128) coupled with the fact that we are retrieving 5000 records (sorted) contribute to the memory bloat when the system isn't able to keep up ? My understanding is that ES will have to sort and return the top 5000 docs on each shard to the coordinating node which will have to sort across 128x5000=640k docs. Our sort field has doc values enabled and it's a long timestamp.

may I ask why you have 128 shards on 3 nodes with 600k doc? What about 1 shard? I also wonder why on earth you are fetching 5k documents instead of using pagination or sorted scroll or search_after?

**Comment by mahdibh** [comment]
> may I ask why you have 128 shards on 3 nodes with 600k doc? What about 1 shard?

This is just a load test environment. The number of shards is artificially high on so that we can determine whether that impact the memory issues we are observing. In my third update, I pointed out that we weren't able to repro this on an index with 8 shards.

> I also wonder why on earth you are fetching 5k documents instead of using pagination or sorted scroll or search_after?

- Pagination would add latency, although we have never tried that. Would doing that be beneficial ? Latency aside, can you explain why that would be more efficient than retrieving 5k in one shot.
- Sorted scrolls would end up being costly due to the scroll context that has to be maintained.
- search_after isn't available in 1.7 afaik.

**Comment by s1monw** [comment]
> This is just a load test environment. The number of shards is artificially high on so that we can determine whether that impact the memory issues we are observing. 

I see - is having 128 shards per node a realistic test? I mean I can put a ton of shards on a node and stuff will likely go bad? Nevertheless, I would be interested in the heapdump of a client node here.

> In my third update, I pointed out that we weren't able to repro this on an index with 8 shards.

sorry missed that....

> Pagination would add latency, although we have never tried that. Would doing that be beneficial ? Latency aside, can you explain why that would be more efficient than retrieving 5k in one shot.

to begin with, why do you need 5k results. Elasticsearch is a top N retrieval engine in the first place where N is small like 10, 20, 30? I doubt the efficiency here, I think you are overestimating latency given that you are fetching 5k * 10kb of json latency becomes a constant factor given the amount of data you are fetching. Also keep in mind elasticsearch has to materialize all that data before sending back ie. it's not streaming. Working with chunks is the way to go.

> Sorted scrolls would end up being costly due to the scroll context that has to be maintained.

what cost you are talking about here? Memory? it holds on to the index reader plus an integer per shard and scroll request. The cost is bascially an int per shard since you need the index reader anyway?

> search_after isn't available in 1.7 afaik.

since you are in a testing phase why don't you go and use 2.x give that 5.x is around the corner?

**Comment by mahdibh** [comment]
One other observation that we made during this is that when the client node runs out of memory, OOM exceptions are swallowed and the node stays up in zombie state holding to its connections to other data nodes. This in turn doesn't release the netty resources on the data nodes leading to a snowball effect that causes the data nodes to also run out of memory.

```
2016-05-12T18:48:01.533Z WARN  [ent_worker][T#7]{New I/O worker #7}] [annel.socket.nio.AbstractNioSelector]: Unexpected exception in the selector loop.
java.lang.OutOfMemoryError: Java heap space
```

If you look at netty's AbstractNioSelector.java code, it's catching throwable, logging it and then moving on. 
If that exception gets bubbled all the way up to the JVM, the process would exit, or at least we could add an OOM handler in the JVM options so that we can kill the process and release the buffers it's holding into the data nodes.

Note that this pattern is also present in other places in the ES code, for example IndicesTTLService.run(), RestActionListener.onFailure() swallow throwable. I'm not sure if those are done on purpose (in the hope that the OOM was momentary?). It seems to me that the best course of action when handling an OOM is either to release memory or to let the exception go up the stack but not to swallow it.

**Comment by mahdibh** [comment]
One more issue we also noted is the following code in MessageChannelHandler
```
            ThrowableObjectInputStream ois = new ThrowableObjectInputStream(buffer, transport.settings().getClassLoader());
            error = (Throwable) ois.readObject();
```

When the data nodes search queues are full, the rejection exception is serialized back to the client node. The deserialization of the EsRejectedExecutionException in ois.readObject() across the different netty worker threads is synchronized in java.lang.ClassLoader.loadClass(ClassLoader.java:357). This blocks all IO worker threads. See two thread dumps here:

https://gist.github.com/mahdibh/859ef5fcb509a95a9dd079baef369546

After this point, it seems that the system gets stuck (probably trying to allocate memory) but never OOMs (or maybe does, but since it's trying to allocate memory to log the OOM, it is never logged).

This is what I see in the logs, but never any OOM. I guess if I wait long enough, it'll likely happen on some other thread that is not trying to catch it.

2016-05-13T13:27:19.416Z INFO  [rch[metabase-es-26][scheduler][T#1]] [monitor.jvm                         ]: [metabase-es-26] [gc][old][1158][256] duration [6.1s], collections [1]/[6.6s], total [6.1s]/[11.5m], memory [3.8gb]->[3.8gb]/[3.9gb], all_pools {[young] [148.1mb]->[174.1mb]/[266.2mb]}{[survivor] [0b]->[0b]/[33.2mb]}{[old] [3.6gb]->[3.6gb]/[3.6gb]}
2016-05-13T13:27:31.022Z INFO  [rch[metabase-es-26][scheduler][T#1]] [monitor.jvm                         ]: [metabase-es-26] [gc][old][1160][258] duration [6.2s], collections [1]/[6.7s], total [6.2s]/[11.7m], memory [3.8gb]->[3.8gb]/[3.9gb], all_pools {[young] [205mb]->[178.8mb]/[266.2mb]}{[survivor] [0b]->[0b]/[33.2mb]}{[old] [3.6gb]->[3.6gb]/[3.6gb]}
2016-05-13T13:27:42.296Z INFO  [rch[metabase-es-26][scheduler][T#1]] [monitor.jvm                         ]: [metabase-es-26] [gc][old][1162][260] duration [6.3s], collections [1]/[6.5s], total [6.3s]/[11.8m], memory [3.9gb]->[3.9gb]/[3.9gb], all_pools {[young] [241.9mb]->[266.2mb]/[266.2mb]}{[survivor] [0b]->[7mb]/[33.2mb]}{[old] [3.6gb]->[3.6gb]/[3.6gb]}

See thread dumps below when it reaches this state
https://gist.github.com/mahdibh/f879362f2c4960e401d2c06385aec5d8

A bit after this, the node logs a ton of the following exceptions:
```

2016-05-13T14:33:08.631Z WARN  [rch[metabase-es-26][generic][T#456]] [search.action                       ]: [metabase-es-26] Failed to send release search context
org.elasticsearch.transport.SendRequestTransportException: [metabase-es-18][inet[/X.X.X.XX:9300]][indices:data/read/search[free_context]]
        at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:286)
        at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:249)
        at org.elasticsearch.search.action.SearchServiceTransportAction.sendFreeContext(SearchServiceTransportAction.java:143)
...
Caused by: org.elasticsearch.transport.NodeNotConnectedException: [metabase-es-18][inet[/X.X.X.X:9300]] Node not connected
        at org.elasticsearch.transport.netty.NettyTransport.nodeChannel(NettyTransport.java:964)
```

Most likely because the node has been kicked out of the cluster and doesn't see the other data nodes.

"Node not connected" is a bit misleading here since it indicates a networking problem reaching the node whereas it's actually just not finding the channel in the node mapping. This could better be characterized as "can't find {node} in the list of connected nodes". I'm mentioning this, because we wasted some time checking our networking when we saw in this issue in production.

**Comment by jasontedor** [comment]
> When the data nodes search queues are full, the rejection exception is serialized back to the client node. The deserialization of the EsRejectedExecutionException in ois.readObject() across the different netty worker threads is synchronized in java.lang.ClassLoader.loadClass(ClassLoader.java:357). This blocks all IO worker threads.

We don't use Java serialization for exceptions anymore. This is not the case in 2.x and master.

**Comment by mahdibh** [comment]
> We don't use Java serialization for exceptions anymore. This is not the case in 2.x and master.
That's great to hear. I would prefer that those do not even get serialized (especially the thread dump). For exceptions that are designed to shed the system from further load, generating them and sending them back to the client should be the cheapest possible. I think a status code indicating a thread pool rejection would be good enough. Not sure what the behavior is in 2.x/master.

2.x is not an option for us at this point because we can't do a rolling upgrade and it'll take us forever to do a full upgrade (and the plan-b if that fails is rather scary ;) ).

**Comment by mahdibh** [comment]
Another thing that could help fix this is to check the netty channel's isWritable() status (see http://normanmaurer.me/presentations/2014-facebook-eng-netty/slides.html#10.0). If the channel isn't writable, the response/request will be queued, hence holding to memory that won't be released until the client/server responds. Under load, data nodes may become slow leading to more queuing on the coordinator side. Also, if the coordinator node is slow receiving, more memory will be held on the data node side. In both cases, OOMs are likely to happen.

The ES code (1.7.5 branch) doesn't have any reference to the isWritable method so there is definitely a risk of running out of memory under these conditions.

**Comment by s1monw** [comment]
just to manage expectations, we won't do another 1.7.x release unless there is a really serious issue coming up. There won't be any features or larger modifications happening to 1.7.x. 
I appreciate your  comments but it's still unclear what exactly is causing the problems on your side? Why is the node going OOM? I also wonder if you are producing an artificial problem here since there haven't been issues reported related to this to my knowledge (I can be wrong). 

There are certainly improvements possible to the network code but I think we should contain them and have a clear purpose. We are more than happy to accept PRs / improvements like the `isWritable()` suggestion!


> 2.x is not an option for us at this point because we can't do a rolling upgrade and it'll take us forever to do a full upgrade (and the plan-b if that fails is rather scary ;) ).

you should really reconsider your strategy - 2.x to 5.x will require a full cluster restart too. I recommend you build some infrastructure to make the move to newer version possible.

**Comment by danielmitterdorfer** [comment]
Here's a summary of what I found out so far:

### Heap dumps

I looked at your heap dumps and the two major heap consumers are:

* `TransportService.clientHandlers` (consuming roughly 1.6GB or 40% of your heap)
* 144 instances of `NioAcceptedSocketChannel` (consuming roughly 2.2GB or 54% of your heap). These instances get created when we accept the network connection. However, if there are not enough workers available they get queued by Netty and the really bad thing is that the queue is an unbounded queue.

This indicates that there are too many open client connections (due to search requests). 

### Reproduction

I tried reproducing the problem with the information that I could find in the ticket. Although I was able to force `EsRejectedExecutionException`s easily, I could not get the system into a state where it OOMs (I am not saying at all that they don't happen; I could just not reproduce it locally).

### Mitigation

However, I think you can do a couple of things to mitigate this situation:

1. First of all, clients should back off when they get `EsRejectedExecutionException` in a response. Since Elasticsearch 2.2, we use an exponential backoff in the Java client (see #14829 and #15513) and you should do something similar if you use another language in your application.
2. You could look whether you can optimize your query. It is fairly complex so maybe you can do something there. It's also not ideal that you do an ascending sort; Elasticsearch is optimized for descending sorts.
3. You should also look into pagination. Do you show 5.000 search results at once to your users?

An upgrade to 2.x could help you for multiple reasons:

1. We have included additional circuit breakers which could maybe help in your situation. The top memory consumers are due to request processing or rather queued requests but it maybe the case that the inflight request breaker can still help in this situation (will be available in release 2.4) although this is not the root cause of the problem here.
2. We have added a query profiler which can help you to find out why your query is slow and help you tune it.

**Comment by mahdibh** [comment]
> just to manage expectations, we won't do another 1.7.x release unless there is a really serious issue coming up

This is a serious issue. If a single node OOMs and gets into this zombie state this is what happens:
1. it completely stalls search requests in the whole cluster. ES becomes completely un-usable.
2. if the node is a coordinator node, it will hold to its connections on data nodes increasing the risk of data nodes also running out of memory. 

Swallowing an OOM is pretty serious too if you talk to any decent java engineer :)  It's a bad practice and leads to all sorts of subtle, hard to diagnose problems. 

Btw, there have been issues reported around this in the past, see #2528 and more specifically #9018

> you should really reconsider your strategy - 2.x to 5.x will require a full cluster restart too. I recommend you build some infrastructure to make the move to newer version possible.

We are already doing that. Building that infrastructure is expensive ! The problem is that the 1.x and 2.x clients aren't compatible, so we have to maintain two client versions with two clusters of different versions while the data is changing underneath. If the clients were kept compatible, that would have saved us a lot of boilerplate work.

**Comment by mahdibh** [comment]
thanks @danielmitterdorfer for taking a look at the dumps.

> if there are not enough workers available they get queued by Netty and the really bad thing is that the queue is an unbounded queue.
Is there a setting to make this queue bounded ?

We have already implemented a backoff strategy to relieve the cluster when requests are being rejected. The result of the queries are actually not showed to users, but used by internal backend systems for other purposes.

What about search results? What happens when a client is too slow to consume a response. Do those also get queued on a queue with an unbounded size ?

**Comment by danielmitterdorfer** [comment]
>> if there are not enough workers available they get queued by Netty and the really bad thing is that the queue is an unbounded queue.

>   Is there a setting to make this queue bounded ?

No, there is no setting in Netty, see [AbstractNioSelector.java#L87](https://github.com/netty/netty/blob/3.10/src/main/java/org/jboss/netty/channel/socket/nio/AbstractNioSelector.java#L87).

> We have already implemented a backoff strategy to relieve the cluster when requests are being rejected.

How do you backoff? Linearly, exponentially, some other strategy? It's also hard to tell without knowing your use case but it seems also that there are quite a lot of clients hitting the cluster (I don't know if you just used 1.000 as an example to produce the condition more quickly or if you really have that many clients hitting it in production). Nevertheless, it seems even with your backoff strategy too many requests reach the server as there are more than one hundred queued connections.

**Comment by jasontedor** [comment]
We agree with you that Elasticsearch should be terminated on any `OutOfMemoryError`. However, the situation is not simple.

First, I think that you're missing that effectively all of the work in Elasticsearch is not done on the main thread. Even without catching `Throwable`, a death of one of these threads due to `OutOfMemoryError` or other `VirtualMachineError`s would (by default) not bring the process down anyway. An option would be to register an `UncaughtExceptionHandler` and terminate the process if this handler is invoked with an `OutOfMemoryError` with the cost of having to unwind all of the code blocks today that are catching `Throwable`.

So maybe a better option is that in later versions of Java (>= 8u92) we will be able to take advantage of `ExitOnOutOfMemoryError` to immediately exit the process if an `OutOfMemoryError` is detected, but that is something that we need to wait until we require Java 9 for.

Even this has complications. For example, it's probably best that when handling a remote request, we catch everything so that we can notify a remote listener before proceeding to tear the process down. This rules out being able to use `ExitOnOutOfMemoryError`.

**Comment by mahdibh** [comment]
> How do you backoff?
We do exponential backoffs.

> I don't know if you just used 1.000 as an example to produce the condition more quickly or if you really have that many clients hitting it in production. 
Since the client is doing async IO, it's actually the number of clients times the number of concurrent queries issued by client (till the IO threads are saturated) that determines the overall load on the system. If you multiply that by the number of shards in the index, the number quickly gets big.

**Comment by mahdibh** [comment]
The core of the issue here is the un-bounded connection queues. That's what netty's `isWritable()` method is for. The handling of the OOM is secondary. It's important that the system protects itself against OOMs. The ES team have been putting a lot of these protections in place as part of the OOM resiliency work and that's great. I think this is another one of these issues that also needs to be added to the list.

Re. swallowing OOM, I think it's just the wrong thing to do regardless of the complications of bubbling up the exception all the way up to the main thread. The JDK documentation for `VirtualMachineError` clearly states: `Thrown to indicate that the Java Virtual Machine is broken or has run out of resources necessary for it to continue operating.`

I would completely understand catching this exception and then for example trying some last resort actions like freeing up memory or closing connections, etc.. However, it should be re-thrown because it's one of those exceptions that you can't just handle and continue to function properly.
The intent of the developer here is to save the system at any cost: catch the OOM and ignore it in the hope that memory will get freed-up somewhere else and the system will get back to working normally. However, that's the best case scenario. Keeping the system up and running in that state may do more harm than good. For example: data corruption (see #12041, #10066) or subtle deadlocks/race conditions that only happen during those low memory conditions.

**Comment by jasontedor** [comment]
Did you see where I said we agree with you? I should add, we had discussions about this awhile ago and already came to the conclusion we should handle it differently than today.

**Comment by danielmitterdorfer** [comment]
@mahdibh I think before we can fix something, we need be able to reproduce the issue (otherwise it's hard to tell we've really fixed the right thing). Therefore, I have started a gist where you can see the reproduction scenario I have created based on your information: https://gist.github.com/danielmitterdorfer/636d1fb1f6a2ec4353a4e0b5d552a391

It would be great if you could fork it and adapt it. One thing I've already noticed: I have not used async IO as I didn't know it before.

**Comment by danielmitterdorfer** [comment]
@mahdibh Did you have any chance to look at the reproduction scenario?

**Comment by mahdibh** [comment]
Sorry @danielmitterdorfer I got pulled into other things. I think this will be hard to reproduce with the python client (I'm assuming it doesn't do any async IO). I'll try to spend some cycles on this next week and will update the ticket.

**Comment by danielmitterdorfer** [comment]
@mahdibh No worries. You can use whatever is most convenient to you, I'm sure we can figure it out then. I just used the Python client when I originally created the scenario because I wasn't aware that you do it async back then (and btw the Python client does not do async IO but Honza has created an async io extension).

**Comment by danielmitterdorfer** [comment]
Due to lack of a reproduction scenario we cannot make progress and I am closing this ticket now. But please feel free to reopen when you've got a reproduction scenario. Btw, @jasontedor started to eliminate the `catch Throwable` blocks in #19231.

**Comment by hexinw** [comment]
Hi,

Can you guys comment if this is a similar instance of OOM exception gets swallowed cases? I have a nested date histogram + cardinality aggregation that would explode the elasticsearch node heap memory. On 5.4.2 ES, I am seeing the OOM exception stack trace thrown properly. I didn't see the OOM stack trace when I upgrade to ES 5.6. Instead Netty exception log is what I see.

[2017-09-12T11:47:27,228][ERROR][o.e.t.n.Netty4Utils      ] fatal error on the network layer
        at org.elasticsearch.transport.netty4.Netty4Utils.maybeDie(Netty4Utils.java:179)
        at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.exceptionCaught(Netty4HttpRequestHandler.java:81)
        at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
        at io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:850)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:364)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at org.elasticsearch.http.netty4.pipelining.HttpPipeliningHandler.channelRead(HttpPipeliningHandler.java:68)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:544)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at java.lang.Thread.run(Thread.java:748)
[2017-09-12T11:47:27,229][WARN ][o.e.m.j.JvmGcMonitorService] [node01-es-dev] [gc][383] overhead, spent [1.3m] collecting in the last [1.3m]
[2017-09-12T11:47:27,227][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [node01-es-dev] fatal error in thread [elasticsearch[node01-es-dev][search][T#3]], exiting
java.lang.OutOfMemoryError: Java heap space
[2017-09-12T11:47:31,496][INFO ][o.e.n.Node               ] [node01-es-dev] initializing ...
[2017-09-12T11:47:31,570][INFO ][o.e.e.NodeEnvironment    ] [node01-es-dev] using [1] data paths, mounts [[/elastic/data (/dev/mapper/elastic-data)]], net usable_space [563.8gb], net total_space [565.7gb], spins? [possibly], types [xfs]
[2017-09-12T11:47:31,570][INFO ][o.e.e.NodeEnvironment    ] [node01-es-dev] heap size [7.9gb], compressed ordinary object pointers [true]
[2017-09-12T11:47:31,598][INFO ][o.e.n.Node               ] [node01-es-dev] node name [node01-es-dev], node ID [MAoYvdPRT02HlKtDD7GQrg]
[2017-09-12T11:47:31,598][INFO ][o.e.n.Node               ] [node01-es-dev] version[5.6.0], pid[16748], build[781a835/2017-09-07T03:09:58.087Z], OS[Linux/3.10.0-514.el7.x86_64/amd64], JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/1.8.0_131/25.131-b12]
[2017-09-12T11:47:31,598][INFO ][o.e.n.Node               ] [node01-es-dev] JVM arguments [-Xms8g, -Xmx8g, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -XX:+DisableExplicitGC, -XX:+AlwaysPreTouch, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Dlog4j.skipJansi=true, -XX:+HeapDumpOnOutOfMemoryError, -Des.path.home=/usr/share/elasticsearch]

**Comment by jasontedor** [comment]
@hexinw The OOM did not get swallowed. Your node stopped and restarted, as expected:

```
[2017-09-12T11:47:27,227][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [node01-es-dev] fatal error in thread [elasticsearch[node01-es-dev][search][T#3]], exiting
java.lang.OutOfMemoryError: Java heap space
[2017-09-12T11:47:31,496][INFO ][o.e.n.Node ] [node01-es-dev] initializing ...
```

**Comment by hexinw** [comment]
So the java.lang.OutOfMemoryError stack trace is not getting logged in time due to node restart? On 5.4.2, I've been able to consistently seeing the stack trace when OOM happens.
